File: android/app/src/main/java/com/cavadvpn/crypto/ReplayFilter.kt
package com.cavadvpn.crypto

import java.nio.ByteBuffer
import java.security.SecureRandom

/** Size in bytes of an encoded PacketHeader. */
const val PACKET_HEADER_SIZE = 20

/** Sliding window radius in seconds for replay detection. */
const val REPLAY_WINDOW_SECS = 90L

/**
 * PacketHeader carries per-packet timestamp and nonce for replay protection.
 * Encoded as 8-byte big-endian timestamp + 12-byte nonce = 20 bytes.
 */
data class PacketHeader(val timestamp: Long, val nonce: ByteArray) {
    init { require(nonce.size == NONCE_SIZE) { "nonce must be $NONCE_SIZE bytes" } }

    fun encode(): ByteArray {
        val buf = ByteBuffer.allocate(PACKET_HEADER_SIZE)
        buf.putLong(timestamp)
        buf.put(nonce)
        return buf.array()
    }

    override fun equals(other: Any?) = other is PacketHeader &&
        timestamp == other.timestamp && nonce.contentEquals(other.nonce)
    override fun hashCode() = 31 * timestamp.hashCode() + nonce.contentHashCode()
}

/** Creates a PacketHeader with the current Unix second and a random nonce. */
fun newPacketHeader(): PacketHeader {
    val nonce = ByteArray(NONCE_SIZE)
    SecureRandom().nextBytes(nonce)
    return PacketHeader(System.currentTimeMillis() / 1000L, nonce)
}

/** Deserialises a PacketHeader from [data] (must be at least 20 bytes). */
fun decodePacketHeader(data: ByteArray): PacketHeader {
    require(data.size >= PACKET_HEADER_SIZE) {
        "data too short: ${data.size} < $PACKET_HEADER_SIZE"
    }
    val buf = ByteBuffer.wrap(data)
    val ts = buf.long
    val nonce = ByteArray(NONCE_SIZE)
    buf.get(nonce)
    return PacketHeader(ts, nonce)
}

/**
 * Thread-safe replay filter with a ±REPLAY_WINDOW_SECS sliding time window.
 * Packets outside the window or with a previously seen (timestamp, nonce) pair are rejected.
 */
class ReplayFilter {
    // second → set of hex-encoded nonces seen in that second
    private val buckets = HashMap<Long, HashSet<String>>()

    /**
     * Returns true if [header] is a fresh packet and records it.
     * Returns false if the packet is a replay or outside the time window.
     */
    @Synchronized
    fun check(header: PacketHeader): Boolean {
        val nowSec = System.currentTimeMillis() / 1000L
        val windowStart = nowSec - REPLAY_WINDOW_SECS
        val windowEnd   = nowSec + REPLAY_WINDOW_SECS

        if (header.timestamp < windowStart || header.timestamp > windowEnd) return false

        // Evict stale buckets
        val iter = buckets.entries.iterator()
        while (iter.hasNext()) {
            if (iter.next().key < windowStart) iter.remove()
        }

        val nonceHex = header.nonce.joinToString("") { "%02x".format(it) }
        val bucket = buckets.getOrPut(header.timestamp) { HashSet() }
        if (bucket.contains(nonceHex)) return false
        bucket.add(nonceHex)
        return true
    }
}

File: ios/Sources/CavadVPNCrypto/ReplayFilter.swift
// ReplayFilter.swift — Sliding window replay protection (±90 seconds)
// Wire-compatible with Go server crypto/replay.go

import Foundation

// MARK: - Constants

/// Size in bytes of an encoded PacketHeader.
public let packetHeaderSize = 20

/// Sliding window half-width in seconds.
public let replayWindowSecs: Int64 = 90

// MARK: - PacketHeader

/// Per-packet timestamp + nonce used for replay protection.
/// Encoding: 8-byte big-endian timestamp + 12-byte nonce = 20 bytes.
public struct PacketHeader: Equatable {
    public let timestamp: Int64  // Unix seconds
    public let nonce: Data       // 12 bytes

    public init(timestamp: Int64, nonce: Data) {
        precondition(nonce.count == nonceSize, "nonce must be \(nonceSize) bytes")
        self.timestamp = timestamp
        self.nonce     = nonce
    }

    /// Encodes the header as 20 bytes (big-endian timestamp + nonce).
    public func encode() -> Data {
        var buf = Data(capacity: packetHeaderSize)
        var ts = timestamp.bigEndian
        withUnsafeBytes(of: &ts) { buf.append(contentsOf: $0) }
        buf.append(nonce)
        return buf
    }
}

/// Creates a PacketHeader with the current Unix second and a random nonce.
public func newPacketHeader() -> PacketHeader {
    let nonce = generateRandomData(count: nonceSize)
    return PacketHeader(timestamp: Int64(Date().timeIntervalSince1970), nonce: nonce)
}

/// Deserialises a PacketHeader from data (must be at least 20 bytes).
public func decodePacketHeader(_ data: Data) throws -> PacketHeader {
    guard data.count >= packetHeaderSize else {
        throw ReplayError.dataTooShort(data.count)
    }
    let tsBytes = data.prefix(8)
    let ts = tsBytes.withUnsafeBytes { $0.load(as: Int64.self).bigEndian }
    let nonce = data[8 ..< 20]
    return PacketHeader(timestamp: ts, nonce: Data(nonce))
}

// MARK: - ReplayFilter

/// Thread-safe replay filter with a ±REPLAY_WINDOW_SECS sliding time window.
/// Packets outside the window or with a previously seen (timestamp, nonce) pair
/// are rejected.
public final class ReplayFilter {
    // second → set of hex-encoded nonces seen in that second
    private var buckets: [Int64: Set<String>] = [:]
    private let lock = NSLock()

    public init() {}

    /// Returns `true` if `header` is a fresh packet (and records it).
    /// Returns `false` if the packet is a replay or outside the time window.
    @discardableResult
    public func check(_ header: PacketHeader) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        let nowSec = Int64(Date().timeIntervalSince1970)
        let windowStart = nowSec - replayWindowSecs
        let windowEnd   = nowSec + replayWindowSecs

        guard header.timestamp >= windowStart, header.timestamp <= windowEnd else {
            return false
        }

        // Evict stale buckets
        buckets = buckets.filter { $0.key >= windowStart }

        let nonceHex = header.nonce.map { String(format: "%02x", $0) }.joined()
        if buckets[header.timestamp]?.contains(nonceHex) == true {
            return false
        }
        buckets[header.timestamp, default: []].insert(nonceHex)
        return true
    }
}

// MARK: - Errors

public enum ReplayError: Error {
    case dataTooShort(Int)
}
